File: coding-interview-linked-lists-advanced/src/main/java/com/example/linkedlist/LinkedListProblems.java
```java
package com.example.linkedlist;

import java.util.HashSet;
import java.util.Stack;

/**
 * This class contains solutions to various common Linked List problems
 * encountered in coding interviews. Each method includes comments
 * explaining the logic, time complexity, and space complexity.
 */
public class LinkedListProblems {

    /**
     * Problem 1: Reverse Linked List
     * Given the head of a singly linked list, reverse the list, and return the reversed list.
     *
     * Approach 1: Iterative
     * We use three pointers: `prev`, `current`, and `nextTemp`.
     * - `prev` keeps track of the previously reversed node (initially null).
     * - `current` points to the node currently being processed (initially head).
     * - `nextTemp` temporarily stores `current.next` before `current.next` is changed.
     * In each iteration, we:
     * 1. Store `current.next` in `nextTemp`.
     * 2. Change `current.next` to `prev`.
     * 3. Move `prev` to `current`.
     * 4. Move `current` to `nextTemp`.
     * This process continues until `current` becomes `null`.
     *
     * Time Complexity: O(N), where N is the number of nodes in the linked list.
     * We iterate through the list exactly once.
     * Space Complexity: O(1).
     * We only use a few extra pointers, regardless of the list size.
     *
     * @param head The head of the linked list to be reversed.
     * @return The head of the reversed linked list.
     */
    public ListNode reverseListIterative(ListNode head) {
        ListNode prev = null;       // Pointer to the previous node (initially null)
        ListNode current = head;    // Pointer to the current node (initially head)

        while (current != null) {
            ListNode nextTemp = current.next; // Store the next node temporarily
            current.next = prev;              // Reverse current node's pointer
            prev = current;                   // Move prev pointer one step forward
            current = nextTemp;               // Move current pointer one step forward
        }
        return prev; // prev will be the new head of the reversed list
    }

    /**
     * Problem 1: Reverse Linked List
     * Approach 2: Recursive
     *
     * The recursive approach works by dividing the problem into smaller subproblems.
     * Base case: If the list is empty or has only one node, it's already reversed.
     * Recursive step:
     * 1. Recursively reverse the rest of the list (`head.next`). This call returns the new head of the reversed sublist.
     * 2. Point `head.next.next` to `head`. This effectively links the current head to the end of the reversed sublist.
     * 3. Set `head.next` to `null` to properly terminate the original head (which is now the tail).
     *
     * Time Complexity: O(N), where N is the number of nodes.
     * Each node is visited once during the recursion.
     * Space Complexity: O(N) due to the recursion stack.
     * In the worst case (a long list), the depth of the recursion can be N.
     *
     * @param head The head of the linked list to be reversed.
     * @return The head of the reversed linked list.
     */
    public ListNode reverseListRecursive(ListNode head) {
        // Base case: If head is null or it's the last node, it's already reversed
        if (head == null || head.next == null) {
            return head;
        }

        // Recursively reverse the rest of the list
        ListNode reversedRest = reverseListRecursive(head.next);

        // After the recursive call returns, `head.next` is the original second node.
        // `head.next.next` should now point back to `head`.
        head.next.next = head;

        // The original head is now the last node, so its next should be null.
        head.next = null;

        return reversedRest; // `reversedRest` is the new head of the entire reversed list
    }


    /**
     * Problem 2: Detect Cycle and Find Start Node
     * Given the head of a linked list, return the node where the cycle begins.
     * If there is no cycle, return null.
     *
     * Approach: Floyd's Tortoise and Hare (Slow and Fast Pointers)
     *
     * Phase 1: Detect the cycle.
     * - Initialize two pointers, `slow` and `fast`, both starting at `head`.
     * - `slow` moves one step at a time (`slow = slow.next`).
     * - `fast` moves two steps at a time (`fast = fast.next.next`).
     * - If there is a cycle, `slow` and `fast` will eventually meet.
     * - If `fast` or `fast.next` becomes `null`, there is no cycle.
     *
     * Phase 2: Find the start of the cycle.
     * - Once `slow` and `fast` meet, reset `slow` to `head`.
     * - Keep `fast` at the meeting point.
     * - Move both `slow` and `fast` one step at a time.
     * - The point where they meet again is the start of the cycle.
     *
     * Mathematical Proof for Phase 2:
     * Let L be the distance from head to cycle start.
     * Let C be the length of the cycle.
     * Let K be the distance from cycle start to meeting point.
     *
     * When slow and fast meet:
     * Distance covered by slow: `Ds = L + K`
     * Distance covered by fast: `Df = L + K + n*C` (where n is number of full cycles fast completed)
     * Also, `Df = 2 * Ds` (because fast moves twice as fast)
     * So, `2 * (L + K) = L + K + n*C`
     * `2L + 2K = L + K + n*C`
     * `L + K = n*C`
     * `L = n*C - K`
     * `L = (n-1)*C + (C - K)`
     * This means that the distance from the head to the cycle start (`L`) is equal to
     * `(n-1)` full cycles plus the remaining distance `(C - K)`.
     * `(C - K)` is the distance from the meeting point back to the cycle start
     * (moving `C-K` steps from the meeting point clockwise brings you to the start of the cycle).
     *
     * So, if we place one pointer at `head` (distance `L` from cycle start) and
     * another pointer at the meeting point (distance `C-K` from cycle start),
     * and move both one step at a time, they will meet at the cycle start.
     *
     * Time Complexity: O(N).
     * In Phase 1, `fast` covers at most `2N` steps, `slow` covers `N` steps.
     * In Phase 2, `slow` and `fast` cover at most `N` steps to meet.
     * Space Complexity: O(1).
     * Uses only a few extra pointers.
     *
     * @param head The head of the linked list.
     * @return The node where the cycle begins, or null if no cycle.
     */
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) {
            return null; // No cycle possible with 0 or 1 node
        }

        ListNode slow = head;
        ListNode fast = head;

        // Phase 1: Detect cycle
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                // Cycle detected, break out of loop to proceed to Phase 2
                break;
            }
        }

        // If fast or fast.next is null, no cycle was found
        if (fast == null || fast.next == null) {
            return null;
        }

        // Phase 2: Find the start of the cycle
        // Reset slow pointer to head
        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }

        return slow; // Both pointers meet at the cycle start node
    }


    /**
     * Problem 3: Merge Two Sorted Lists
     * You are given the heads of two sorted linked lists, list1 and list2.
     * Merge the two lists into a single sorted list.
     * The list should be made by splicing together the nodes of the first two lists.
     * Return the head of the merged linked list.
     *
     * Approach: Iterative with a Dummy Node
     * - Create a dummy node to simplify the logic for handling the head of the merged list.
     * - Use a `current` pointer to build the new list.
     * - Compare the values of the nodes pointed to by `list1` and `list2`.
     * - Append the smaller node to `current.next` and advance that list's pointer.
     * - Advance `current` to its newly appended node.
     * - After one list is exhausted, append the remaining nodes of the other list.
     *
     * Time Complexity: O(M + N), where M and N are the lengths of list1 and list2.
     * We iterate through both lists once.
     * Space Complexity: O(1).
     * We only use a few extra pointers; no new nodes are created (we reuse existing nodes).
     *
     * @param list1 The head of the first sorted linked list.
     * @param list2 The head of the second sorted linked list.
     * @return The head of the merged sorted linked list.
     */
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        // Create a dummy node to act as the placeholder for the head of the merged list.
        // This simplifies handling the first node.
        ListNode dummyHead = new ListNode(0);
        ListNode current = dummyHead; // Pointer to build the new list

        // Iterate while both lists have nodes remaining
        while (list1 != null && list2 != null) {
            if (list1.val <= list2.val) {
                current.next = list1; // Append node from list1
                list1 = list1.next;   // Move list1 pointer forward
            } else {
                current.next = list2; // Append node from list2
                list2 = list2.next;   // Move list2 pointer forward
            }
            current = current.next; // Move current pointer forward in the merged list
        }

        // If one list is exhausted, append the remaining nodes of the other list
        if (list1 != null) {
            current.next = list1;
        } else {
            current.next = list2;
        }

        return dummyHead.next; // The actual head is after the dummy node
    }

    /**
     * Problem 4: Remove Nth Node From End of List
     * Given the head of a linked list, remove the n-th node from the end of the list and return its head.
     *
     * Approach: Two Pointers (Fast and Slow)
     * - Use a dummy node to handle edge cases like removing the head of the list.
     * - Initialize two pointers, `fast` and `slow`, both starting at the dummy node.
     * - Move `fast` `n` steps ahead. Now `fast` is `n` nodes ahead of `slow`.
     * - Move both `fast` and `slow` one step at a time until `fast` reaches the end of the list (`fast == null`).
     * - At this point, `slow` will be pointing to the node *before* the n-th node from the end.
     * - To remove the n-th node, update `slow.next = slow.next.next`.
     *
     * Time Complexity: O(N), where N is the number of nodes in the linked list.
     * We iterate through the list at most twice (once for fast to move `n` steps, then both move till end).
     * Space Complexity: O(1).
     * Uses only a few extra pointers.
     *
     * Edge Cases:
     * - `n` is equal to the length of the list (remove head).
     * - `n` is 1 (remove tail).
     * - List has only one node.
     * The dummy node handles the "remove head" case elegantly.
     *
     * @param head The head of the linked list.
     * @param n The position from the end of the list to remove.
     * @return The head of the modified linked list.
     */
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // Create a dummy node to simplify edge cases, especially when the head needs to be removed.
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode fast = dummy;
        ListNode slow = dummy;

        // Move fast pointer n steps ahead.
        // This creates a gap of n nodes between fast and slow.
        for (int i = 0; i < n; i++) {
            // If n is larger than the list length, or list is null (shouldn't happen per problem constraints, but good defensive check)
            if (fast == null) return head; // Or throw an error depending on problem spec
            fast = fast.next;
        }

        // Move both fast and slow pointers until fast reaches the end of the list.
        // When fast reaches null, slow will be pointing to the node *before* the Nth node from the end.
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }

        // Now slow.next is the Nth node from the end. Remove it.
        slow.next = slow.next.next;

        return dummy.next; // Return the new head (which might be the original head or null if it was removed)
    }

    /**
     * Problem 5: Reorder List
     * Given the head of a singly linked list. Reorder the list such that:
     * L0 → LN → L1 → LN-1 → L2 → LN-2 → ...
     * You may not modify the values in the list's nodes. Only nodes themselves may be changed.
     *
     * Approach:
     * This problem can be broken down into three main steps:
     * 1. Find the middle of the linked list.
     * 2. Reverse the second half of the linked list.
     * 3. Merge the first half and the reversed second half alternately.
     *
     * Step 1: Find the middle of the list.
     * Use slow and fast pointers. When fast reaches the end, slow is at the middle.
     * If the list has an odd number of nodes, slow will be the exact middle.
     * If even, slow will be the first of the two middle nodes.
     *
     * Step 2: Reverse the second half.
     * Disconnect the first half from the second half by setting `slow.next = null`.
     * Then, use the `reverseListIterative` method (or a similar one) to reverse the second half.
     *
     * Step 3: Merge the two halves.
     * Alternately take nodes from the first half and the reversed second half.
     *
     * Time Complexity: O(N), where N is the number of nodes in the linked list.
     * - Finding middle: O(N/2) = O(N)
     * - Reversing second half: O(N/2) = O(N)
     * - Merging: O(N/2) = O(N)
     * Total: O(N)
     * Space Complexity: O(1).
     * We only use a few extra pointers. No recursion stack for reversal if using iterative approach.
     *
     * @param head The head of the linked list to be reordered.
     */
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return; // No reordering needed for empty or single-node lists
        }

        // Step 1: Find the middle of the list
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        // At this point, slow is at the middle.
        // For 1->2->3->4->5, slow is 3.
        // For 1->2->3->4, slow is 3.

        // Step 2: Split the list into two halves and reverse the second half
        ListNode firstHalfHead = head;
        ListNode secondHalfHead = slow.next; // Start of the second half
        slow.next = null; // Disconnect the two halves

        secondHalfHead = reverseListIterative(secondHalfHead); // Use the iterative reversal from above

        // Step 3: Merge the two halves alternately
        // first half: L0 -> L1 -> L2 -> ...
        // second half: LN -> LN-1 -> LN-2 -> ...
        // Result: L0 -> LN -> L1 -> LN-1 -> L2 -> LN-2 -> ...
        ListNode p1 = firstHalfHead;
        ListNode p2 = secondHalfHead;

        while (p1 != null && p2 != null) {
            ListNode nextP1 = p1.next; // Store next of p1
            ListNode nextP2 = p2.next; // Store next of p2

            p1.next = p2;       // Link p1 to p2
            p2.next = nextP1;   // Link p2 to original next of p1

            p1 = nextP1; // Move p1 to its original next
            p2 = nextP2; // Move p2 to its original next
        }
        // If p1 is not null, it means the original list had an odd number of nodes,
        // and p1 is the last node, which is already correctly linked.
        // If p2 is not null, it means there was an issue, which shouldn't happen if list lengths are correctly handled.
    }
}
```

File: coding-interview-linked-lists-enhanced/linked_list_project/main_algorithms/linked_list_problems.py
```python
from typing import Optional, List
# Assuming ListNode and utilities are available via package structure or relative import
from utils.linked_list_utils import ListNode, list_to_linked_list, linked_list_to_list

class LinkedListProblems:
    """
    A collection of common Linked List problems with optimal solutions.
    """

    @staticmethod
    def reverse_list_iterative(head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Problem 1: Reverse a Singly Linked List (Iterative Approach)

        Given the head of a singly linked list, reverse the list, and return the reversed list.

        Example:
            Input: head = [1,2,3,4,5]
            Output: [5,4,3,2,1]

        Approach: Iterative
        We use three pointers: `prev`, `curr`, and `next_temp`.
        - `prev` keeps track of the previous node (initially None).
        - `curr` is the current node we are processing (initially `head`).
        - `next_temp` temporarily stores the next node before `curr.next` is changed.

        Algorithm:
        1. Initialize `prev = None` and `curr = head`.
        2. While `curr` is not `None`:
           a. Store the next node: `next_temp = curr.next`.
           b. Reverse the current node's pointer: `curr.next = prev`.
           c. Move `prev` to `curr`: `prev = curr`.
           d. Move `curr` to `next_temp`: `curr = next_temp`.
        3. After the loop, `prev` will be the new head of the reversed list.

        Time Complexity: O(N), where N is the number of nodes in the linked list.
                         We visit each node exactly once.
        Space Complexity: O(1). We only use a few extra pointers, independent of input size.
        """
        prev: Optional[ListNode] = None
        curr: Optional[ListNode] = head

        while curr:
            next_temp = curr.next  # Store next node
            curr.next = prev       # Reverse current node's pointer
            prev = curr            # Move prev to current node
            curr = next_temp       # Move current to next node

        return prev # prev will be the new head

    @staticmethod
    def reverse_list_recursive(head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Problem 1: Reverse a Singly Linked List (Recursive Approach)

        Given the head of a singly linked list, reverse the list, and return the reversed list.

        Approach: Recursive
        The idea is to recursively reverse the sublist starting from `head.next`,
        and then fix the pointers for the current `head` node.

        Algorithm:
        1. Base case: If `head` is `None` or `head.next` is `None` (single node), it's already reversed.
           Return `head`.
        2. Recursive step:
           a. Recursively call `reverse_list_recursive` on `head.next`. Let the result be `new_head`.
              `new_head` will be the actual head of the fully reversed list (e.g., if input is 1->2->3,
              after reversing 2->3 to 3->2, `new_head` is 3).
           b. Now, `head.next` points to the node that was originally next to `head`.
              This node (let's call it `second_node`) now has its `next` pointer pointing to `None`
              (because its original `next` was reversed to point to `head.next.next`).
              We need to make `second_node.next` point back to `head`.
              So, `head.next.next = head`.
           c. Set `head.next = None` to break the original link and ensure `head` becomes the tail.
        3. Return `new_head`.

        Time Complexity: O(N), where N is the number of nodes. Each node is visited once.
        Space Complexity: O(N) due to the recursion call stack. In the worst case (a long list),
                          it can lead to stack overflow for very large N.
        """
        # Base case: empty list or single node list
        if not head or not head.next:
            return head

        # Recursively reverse the rest of the list
        new_head = LinkedListProblems.reverse_list_recursive(head.next)

        # After the recursive call returns, 'head.next' is the head of the reversed sublist.
        # For example, if original was 1->2->3->4, and new_head is 4->3->2,
        # then head is 1, head.next is 2. The sublist 2->3->4 was reversed to 4->3->2.
        # Now, we need to connect 1 to the end of the reversed sublist (which is 2).
        head.next.next = head # Make the original next node point back to head
        head.next = None      # Disconnect head from its original next (it's now the tail)

        return new_head # The actual head of the fully reversed list

    @staticmethod
    def has_cycle(head: Optional[ListNode]) -> bool:
        """
        Problem 2: Linked List Cycle

        Given the head of a linked list, return true if the list has a cycle in it.
        There is a cycle in a linked list if there is some node in the list that can be reached
        again by continuously following the next pointer.

        Approach: Floyd's Tortoise and Hare (Fast and Slow Pointers)
        This algorithm uses two pointers, a 'slow' pointer that moves one step at a time,
        and a 'fast' pointer that moves two steps at a time.
        If there is a cycle, the fast pointer will eventually catch up to the slow pointer
        (they will meet at some node within the cycle). If there is no cycle, the fast pointer
        will reach the end of the list (None).

        Algorithm:
        1. Initialize `slow = head` and `fast = head`.
        2. While `fast` and `fast.next` are not `None`:
           a. Move `slow` one step: `slow = slow.next`.
           b. Move `fast` two steps: `fast = fast.next.next`.
           c. If `slow == fast`, a cycle is detected. Return `True`.
        3. If the loop finishes, it means `fast` (or `fast.next`) became `None`, so there's no cycle.
           Return `False`.

        Time Complexity: O(N). In the worst case (no cycle or a very long cycle),
                         the fast pointer traverses the list. Each step of fast covers 2 nodes,
                         and slow covers 1 node. So, total work is proportional to N.
        Space Complexity: O(1). Only two pointers are used.
        """
        if not head:
            return False

        slow: Optional[ListNode] = head
        fast: Optional[ListNode] = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        
        return False

    @staticmethod
    def detect_cycle_start(head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Problem 2 Variation: Find the starting node of the cycle.

        If there is a cycle, return the node where the cycle begins.
        If there is no cycle, return None.

        Approach: Floyd's Tortoise and Hare with a second pass.
        1. Use the standard Floyd's algorithm to detect a cycle. If no cycle, return None.
        2. If a cycle is detected (slow == fast), reset one pointer (e.g., `slow_again = head`).
        3. Move both `slow_again` and `fast` (the meeting point pointer) one step at a time.
        4. The node where they meet is the start of the cycle.

        Proof (Intuition):
        - Let `D` be the distance from head to the start of the cycle.
        - Let `L` be the length of the cycle.
        - Let `K` be the distance from the cycle start to the meeting point.

        When `slow` and `fast` meet:
        - `slow` has traveled `D + K` steps.
        - `fast` has traveled `D + K + m*L` steps (where `m` is some integer for cycle traversals).
        - Since `fast` moves twice as fast as `slow`: `2 * (D + K) = D + K + m*L`
        - Simplifying: `D + K = m*L`
        - Rearranging: `D = m*L - K` or `D = (m-1)*L + (L-K)`

        This means that the distance from the head to the cycle start (`D`) is equal to
        the distance from the meeting point back to the cycle start (`L-K`, potentially plus full cycles).
        So, if we put one pointer at `head` and one at `meeting_point`, and move them one step at a time,
        they will meet exactly at the cycle's starting node.

        Time Complexity: O(N). Two passes over the list in the worst case (first to detect, second to find start).
        Space Complexity: O(1). Only a few pointers are used.
        """
        if not head:
            return None

        slow: Optional[ListNode] = head
        fast: Optional[ListNode] = head
        
        # Step 1: Detect if a cycle exists
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                # Cycle detected
                break
        else:
            # No cycle found (fast or fast.next became None)
            return None

        # Step 2: Find the starting node of the cycle
        # If a cycle exists, 'slow' and 'fast' met.
        # Now, move one pointer to the head, and both move one step at a time.
        # They will meet at the start of the cycle.
        pointer1: Optional[ListNode] = head
        pointer2: Optional[ListNode] = slow # or fast, doesn't matter, they are at the same point

        while pointer1 != pointer2:
            pointer1 = pointer1.next
            pointer2 = pointer2.next
        
        return pointer1 # This is the start of the cycle


    @staticmethod
    def merge_two_lists_iterative(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        """
        Problem 3: Merge Two Sorted Lists (Iterative Approach)

        Merge two sorted linked lists and return it as a new sorted list.
        The new list should be made by splicing together the nodes of the first two lists.

        Example:
            Input: l1 = [1,2,4], l2 = [1,3,4]
            Output: [1,1,2,3,4,4]

        Approach: Iterative
        Use a dummy node to simplify the logic of adding to the merged list and handling the head.
        Maintain a `current` pointer to the tail of the merged list.

        Algorithm:
        1. Create a `dummy_head` node (e.g., with value 0 or -1).
        2. Initialize `current = dummy_head`.
        3. While both `l1` and `l2` are not `None`:
           a. Compare `l1.val` and `l2.val`.
           b. Append the smaller node to `current.next`.
           c. Move the corresponding list pointer forward (e.g., if `l1.val` was smaller, `l1 = l1.next`).
           d. Move `current` to the newly added node (`current = current.next`).
        4. After the loop, one of the lists might still have remaining elements.
           Append the rest of the non-empty list to `current.next`.
        5. The merged list starts from `dummy_head.next`.

        Time Complexity: O(N + M), where N and M are the lengths of l1 and l2 respectively.
                         We visit each node from both lists exactly once.
        Space Complexity: O(1) if we are modifying the existing nodes' `next` pointers.
                          (The space for the new dummy node is constant).
        """
        dummy_head = ListNode(0) # Sentinel node
        current = dummy_head

        while l1 and l2:
            if l1.val <= l2.val:
                current.next = l1
                l1 = l1.next
            else:
                current.next = l2
                l2 = l2.next
            current = current.next
        
        # Append the remaining list (if any)
        if l1:
            current.next = l1
        elif l2:
            current.next = l2
        
        return dummy_head.next

    @staticmethod
    def merge_two_lists_recursive(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        """
        Problem 3: Merge Two Sorted Lists (Recursive Approach)

        Approach: Recursive
        The base cases are when one of the lists is empty.
        In the recursive step, compare the heads of `l1` and `l2`.
        The smaller node becomes the head of the merged list, and its `next` pointer
        is set to the result of merging the rest of its list with the other list.

        Algorithm:
        1. Base cases:
           a. If `l1` is `None`, return `l2`.
           b. If `l2` is `None`, return `l1`.
        2. Recursive step:
           a. If `l1.val <= l2.val`:
              `l1.next = merge_two_lists_recursive(l1.next, l2)`
              Return `l1`.
           b. Else (`l2.val < l1.val`):
              `l2.next = merge_two_lists_recursive(l1, l2.next)`
              Return `l2`.

        Time Complexity: O(N + M). Each recursive call performs constant work and reduces the total
                         number of nodes to process. There will be N+M calls.
        Space Complexity: O(N + M) due to the recursion call stack. In the worst case (e.g., one list much
                          longer than the other, or alternating elements), the depth of recursion
                          can be N+M.
        """
        # Base cases
        if not l1:
            return l2
        if not l2:
            return l1

        # Recursive step
        if l1.val <= l2.val:
            l1.next = LinkedListProblems.merge_two_lists_recursive(l1.next, l2)
            return l1
        else:
            l2.next = LinkedListProblems.merge_two_lists_recursive(l1, l2.next)
            return l2


    @staticmethod
    def remove_nth_from_end(head: Optional[ListNode], n: int) -> Optional[ListNode]:
        """
        Problem 4: Remove Nth Node From End of List

        Given the head of a linked list, remove the nth node from the end of the list and return its head.

        Example:
            Input: head = [1,2,3,4,5], n = 2
            Output: [1,2,3,5]
            (Removes 4)

            Input: head = [1], n = 1
            Output: []

            Input: head = [1,2], n = 2
            Output: [2]
            (Removes 1)

        Approach: Two Pointers (Single Pass)
        We use two pointers, `slow` and `fast`, separated by `n` nodes.
        When `fast` reaches the end of the list, `slow` will be at the node *before* the Nth node from the end.
        This allows `slow` to directly modify `slow.next` to remove the target node.

        Algorithm:
        1. Create a `dummy_head` node and point its `next` to `head`. This simplifies edge cases
           like removing the head node.
        2. Initialize `slow = dummy_head` and `fast = dummy_head`.
        3. Move `fast` `n+1` steps forward. This creates a gap of `n` nodes between `slow` and `fast`,
           and `fast` is one node ahead of where it would be if the gap was just `n`.
           The reason for `n+1` steps is to ensure `slow` stops *before* the node to be removed.
        4. While `fast` is not `None`:
           a. Move both `slow` and `fast` one step forward.
        5. When `fast` reaches `None`, `slow` will be pointing to the node directly preceding
           the node to be removed.
        6. Remove the node: `slow.next = slow.next.next`.
        7. Return `dummy_head.next`.

        Time Complexity: O(N), where N is the number of nodes. Both pointers traverse the list once.
        Space Complexity: O(1). Only a few extra pointers are used.
        """
        dummy_head = ListNode(0)
        dummy_head.next = head

        slow: Optional[ListNode] = dummy_head
        fast: Optional[ListNode] = dummy_head

        # Move fast n+1 steps ahead of slow
        # This makes the gap between slow and fast (n nodes + the node to be removed)
        # When fast reaches end, slow will be at node *before* the one to be removed
        for _ in range(n + 1):
            if fast: # Ensure fast doesn't become None prematurely
                fast = fast.next
            else:
                # This case implies n is larger than the list length, or head was None
                # LeetCode problem constraints usually guarantee n is valid, but defensive check
                return head # Or raise an error, depending on spec

        # Move both pointers until fast reaches the end
        while fast:
            slow = slow.next
            fast = fast.next
        
        # Now, slow is pointing to the node right before the one to be removed
        # The node to remove is slow.next
        if slow and slow.next:
            slow.next = slow.next.next
        
        return dummy_head.next

    @staticmethod
    def add_two_numbers(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        """
        Problem 5: Add Two Numbers

        You are given two non-empty linked lists representing two non-negative integers.
        The digits are stored in reverse order, and each of their nodes contains a single digit.
        Add the two numbers and return the sum as a linked list.
        You may assume the two numbers do not contain any leading zero, except the number 0 itself.

        Example:
            Input: l1 = [2,4,3], l2 = [5,6,4]  (Represents 342 + 465)
            Output: [7,0,8]                    (Represents 807)

        Approach: Iterative with Carry
        Simulate manual addition process, digit by digit, from right to left (which is head to tail in reverse order).

        Algorithm:
        1. Create a `dummy_head` node (e.g., with value 0). This will point to the head of the result list.
        2. Initialize `current = dummy_head` and `carry = 0`.
        3. Loop while `l1` is not `None` OR `l2` is not `None` OR `carry` is not 0:
           a. Get the values of the current digits (or 0 if a list is exhausted):
              `val1 = l1.val if l1 else 0`
              `val2 = l2.val if l2 else 0`
           b. Calculate the sum: `total_sum = val1 + val2 + carry`.
           c. Update `carry`: `carry = total_sum // 10`.
           d. Get the digit for the new node: `digit = total_sum % 10`.
           e. Create a new node with `digit` and append it to `current.next`.
           f. Move `current` to the new node: `current = current.next`.
           g. Advance `l1` and `l2` if they are not `None`.
        4. Return `dummy_head.next`.

        Time Complexity: O(max(N, M)), where N and M are the lengths of l1 and l2.
                         We iterate through the longer list at most once.
        Space Complexity: O(max(N, M)). A new linked list is created to store the sum,
                          which will have at most max(N, M) + 1 nodes.
        """
        dummy_head = ListNode(0)
        current = dummy_head
        carry = 0

        while l1 or l2 or carry: # Continue as long as there are digits or a carry
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0

            # Calculate sum and new carry
            total_sum = val1 + val2 + carry
            carry = total_sum // 10
            digit = total_sum % 10

            # Create new node and append
            current.next = ListNode(digit)
            current = current.next

            # Move to the next nodes in l1 and l2
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
        
        return dummy_head.next
```